src/main/host/process.rs

```diff
@@ -440,24 +443,67 @@ impl Process {
             .chain(std::iter::once(std::ptr::null()))
             .collect();
 
-        let cthread = unsafe { main_thread.cthread() };
+        unsafe {
+            cshadow::thread_run(
+                main_thread.cthread(),
+                self.plugin_path.as_ptr(),
+                argv_ptrs.as_ptr(),
+                envv_ptrs.as_ptr(),
+                self.working_dir.as_ptr(),
+                match &self.strace_logging {
+                    Some(x) => x.file.borrow().as_raw_fd(),
+                    None => -1,
+                },
+            )
+        };
+
+        let native_pid =
+            Pid::from_raw(unsafe { cshadow::thread_getNativePid(main_thread.cthread()) });
+        self.native_pid.set(Some(native_pid));
+        *self.memory_manager.borrow_mut() = Some(unsafe { MemoryManager::new(native_pid) });
+
+        #[cfg(feature = "perf_timers")]
+        {
+            let elapsed = self.cpu_delay_timer.borrow_mut().stop();
+            info!("process '{}' started in {:?}", self.name(), elapsed);
+        }
+        #[cfg(not(feature = "perf_timers"))]
+        info!("process '{}' started", self.name());
+
+        Worker::clear_active_thread();
+        Worker::clear_active_process();
+
+        if self.pause_for_debugging {
+            // will block until logger output has been flushed
+            // there is a race condition where other threads may log between the fprintf() and raise()
+            // below, but it should be rare
+            log::logger().flush();
+
+            // Use a single `eprintln` to ensure an atomic write to stderr.
+            eprintln!(
+                "\
+              \n** Pausing with SIGTSTP to enable debugger attachment to managed process\
```

Comment:
This is atomic in the context of rust's locked stderr, but it will still make multiple `write()` syscalls, so I think it would be better to `format!()` the string first and then print it, which will hopefully require only a single syscall. Then if anything else such as C code in another thread writes to stderr, it should hopefully not be broken up. (I'm not sure that the original C code was using a single syscall either.)

Reply:
Done